File: core/src/main/java/com/ortin/gesturetranslator/core/repository/RecognizeCoordinateRepositoryImpl.java
package com.ortin.gesturetranslator.core.repository;

import android.util.Log;

import com.ortin.gesturetranslator.core.managers.model_coordinate.ModelCoordinateManager;
import com.ortin.gesturetranslator.core.managers.model_coordinate.models.ModelCoordinateArray;
import com.ortin.gesturetranslator.core.managers.model_coordinate.models.ModelCoordinateClassificationArray;
import com.ortin.gesturetranslator.domain.models.CoordinateClassification;
import com.ortin.gesturetranslator.domain.models.HandDetected;
import com.ortin.gesturetranslator.domain.repository.RecognizeCoordinateRepository;

public class RecognizeCoordinateRepositoryImpl implements RecognizeCoordinateRepository {
    private ModelCoordinateManager modelCoordinateManager;

    public RecognizeCoordinateRepositoryImpl(ModelCoordinateManager modelCoordinateManager) {
        this.modelCoordinateManager = modelCoordinateManager;
    }

    @Override
    public CoordinateClassification recognise(HandDetected handDetected) {
        return mapToCoreCoordinateClassification(modelCoordinateManager.recognise(mapToCoreCoordinate(handDetected)));
    }


    // Правила перевода для связи domain и core модулей

    private ModelCoordinateArray mapToCoreCoordinate(HandDetected handDetected) {
        float[] oldCoord = handDetected.getCoordinates();
        double[] coord = new double[oldCoord.length];

        for (int i = 0; i < oldCoord.length; i++) {
            coord[i] = (double) oldCoord[i];
        }

        double mainX = coord[0];
        double mainY = coord[1];

        double minX = 10, minY = 10;

        for (int i = 0; i < coord.length; i += 2) {
            coord[i] = coord[i] - mainX;
            coord[i + 1] = coord[i + 1] - mainY;
            coord[i + 1] = Math.abs(coord[i + 1]);
            minX = Math.min(minX, coord[i]);
            minY = Math.min(minY, coord[i + 1]);
        }

        double offsetX = minX < 0 ? Math.abs(minX) : 0;
        double offsetY = minY < 0 ? Math.abs(minY) : 0;

        double maxWidth = 1;
        double maxHeight = 1;

        double maxX = 0, maxY = 0;

        for (int i = 0; i < coord.length; i += 2) {
            coord[i] += offsetX;
            coord[i + 1] += offsetY;
            maxX = Math.max(maxX, coord[i]);
            maxY = Math.max(maxY, coord[i + 1]);
        }

        double coefY = maxY != 0 ? Math.abs(maxHeight / maxY) : 0;
        double coefX = maxX != 0 ? Math.abs(maxWidth / maxX) : 0;

        for (int i = 0; i < coord.length; i += 2) {
            coord[i] *= coefX;
            coord[i + 1] *= coefY;
        }

        return new ModelCoordinateArray(coord);
    }

    private CoordinateClassification mapToCoreCoordinateClassification(ModelCoordinateClassificationArray modelCoordinateClassificationArray) {
        return new CoordinateClassification(modelCoordinateClassificationArray.getLabel(), modelCoordinateClassificationArray.getPercent());
    }
}

File: core/src/main/kotlin/com/ortin/gesturetranslator/core/repository/RecognizeCoordinateRepositoryImpl.kt
package com.ortin.gesturetranslator.core.repository

import com.ortin.gesturetranslator.core.managers.model_coordinate.ModelCoordinateManager
import com.ortin.gesturetranslator.core.managers.model_coordinate.models.ModelCoordinateArray
import com.ortin.gesturetranslator.core.managers.model_coordinate.models.ModelCoordinateClassificationArray
import com.ortin.gesturetranslator.domain.models.CoordinateClassification
import com.ortin.gesturetranslator.domain.models.ImageDetected
import com.ortin.gesturetranslator.domain.repository.RecognizeCoordinateRepository
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.abs

@Singleton
class RecognizeCoordinateRepositoryImpl @Inject constructor(private val modelCoordinateManager: ModelCoordinateManager) :
    RecognizeCoordinateRepository {
    override fun recognise(imageDetected: ImageDetected): CoordinateClassification =
        mapToCoreCoordinateClassification(
            modelCoordinateManager.recognise(
                mapToCoreCoordinate(
                    imageDetected
                )
            )
        )

    // Правила перевода для связи domain и core модулей
    private fun mapToCoreCoordinate(imageDetected: ImageDetected): ModelCoordinateArray {
        val oldCoord = imageDetected.coordinates
        val coord = List(oldCoord.size) { index -> oldCoord[index].toDouble() }.toMutableList()

        val mainX = coord[0]
        val mainY = coord[1]
        var minX = 10.0
        var minY = 10.0

        for (i in coord.indices step 2) {
            coord[i] = coord[i] - mainX
            coord[i + 1] = coord[i + 1] - mainY
            coord[i + 1] = abs(coord[i + 1])
            minX = minX.coerceAtMost(coord[i])
            minY = minY.coerceAtMost(coord[i + 1])
        }

        val offsetX = if (minX < 0) abs(minX) else 0.0
        val offsetY = if (minY < 0) abs(minY) else 0.0
        val maxWidth = 1.0
        val maxHeight = 1.0
        var maxX = 0.0
        var maxY = 0.0

        for (i in coord.indices step 2) {
            coord[i] += offsetX
            coord[i + 1] += offsetY
            maxX = maxX.coerceAtLeast(coord[i])
            maxY = maxY.coerceAtLeast(coord[i + 1])
        }

        val coefY = if (maxY != 0.0) abs(maxHeight / maxY) else 0.0
        val coefX = if (maxX != 0.0) abs(maxWidth / maxX) else 0.0

        for (i in coord.indices step 2) {
            coord[i] *= coefX
            coord[i + 1] *= coefY
        }

        return ModelCoordinateArray(coord)
    }

    private fun mapToCoreCoordinateClassification(modelCoordinateClassificationArray: ModelCoordinateClassificationArray): CoordinateClassification =
        CoordinateClassification(
            modelCoordinateClassificationArray.label,
            modelCoordinateClassificationArray.percent
        )
}
